src/storage/src/source/generator/upsert.rs

```diff
@@ -0,0 +1,556 @@
+        let mut updates_buffer = Vec::new();
+
+        // snapshotting
+        let mut upper_offset = if snapshotting {
+            if stats_worker {
+                stats_output
+                    .give(
+                        &stats_cap,
+                        ProgressStatisticsUpdate::SteadyState {
+                            offset_known: upsert.snapshot_rounds,
+                            offset_committed: 0,
+                        },
+                    )
+                    .await;
+            };
+
+            // Downgrade to the
+            progress_cap.downgrade(&MzOffset::from(upsert.snapshot_rounds));
+
+            let mut emitted = 0;
+            stats_output
+                .give(
+                    &stats_cap,
+                    ProgressStatisticsUpdate::Snapshot {
+                        records_known: local_snapshot_size,
+                        records_staged: emitted,
+                    },
+                )
+                .await;
+            while local_partitions.values().any(|si| !si.finished()) {
+                for sp in local_partitions.values_mut() {
+                    updates_buffer.clear();
+                    emitted += sp.produce_batch(&mut updates_buffer, &mut value_buffer);
+                    data_output.give_container(&cap, &mut updates_buffer).await;
+
+                    stats_output
+                        .give(
+                            &stats_cap,
+                            ProgressStatisticsUpdate::Snapshot {
+                                records_known: local_snapshot_size,
+                                records_staged: emitted,
+                            },
+                        )
+                        .await;
+                }
+                tokio::task::yield_now().await;
+            }
+
+            cap.downgrade(&MzOffset::from(upsert.snapshot_rounds));
+            upsert.snapshot_rounds
+        } else {
+            cap.downgrade(&resume_offset);
+            progress_cap.downgrade(&resume_offset);
+            resume_offset.offset
+        };
+
+        if upsert.update_rate.is_some() || upsert.quick_rounds > 0 {
+            let mut local_partitions: BTreeMap<_, _> = (0..upsert.partitions)
+                .into_iter()
+                .filter_map(|p| {
+                    config.responsible_for(p).then(|| {
+                        (
+                            p,
+                            UpdateProducer::new(
+                                p,
+                                upsert.partitions,
+                                upsert.keys,
+                                upsert.snapshot_rounds,
+                                upsert.quick_rounds,
+                                upsert.batch_size,
+                                upsert.seed,
+                                upper_offset,
+                            ),
+                        )
+                    })
+                })
+                .collect();
```

Comment:
if we have no `local_partitions` we can early return, right? otherwise it looks like we might hot-loop below (the `yield_now` is preventing that but it might be unnecessary)

Reply:
we will loop on the `interval.tick()`, but I can break out I think!